### PR TITLE
adds serializer for Celo CIP 42 transactions

### PR DESCRIPTION
**Pull request by aaronmgdr**
followup from #691

A custom Serializer supporting Celo CIP-42 (fee currency) Transactions.

---

### Detailed summary
- Added serializer for Celo CIP-42 transactions.
- Imported `celoSerializers` in `src/chains/index.ts`.
- Added `serializeTransaction` function in `src/chains/serializers/celo.ts`.
- Added utility functions and types related to CIP-42 transactions in `src/chains/serializers/celo.ts`.
- Added tests for CIP-42 transaction serialization in `src/chains/serializers/celo.test.ts`.

> The following files were skipped due to too many changes: `src/chains/serializers/celo.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/chains/serializers/celo.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    gatewayFee ? toHex(gatewayFee) : '0x',
+    to ?? '0x',
+    value ? toHex(value) : '0x',
+    data ?? '0x',
+    serializeAccessList(accessList),
+  ]
+
+  if (signature) {
+    serializedTransaction.push(
+      signature.v === 27n ? '0x' : toHex(1), // yParity
```
Comment (aaronmgdr) [inline src/chains/serializers/celo.ts:91]: need to confirm this is correct

File: src/chains/serializers/celo.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    gatewayFeeRecipient,
+  } = transaction
+  // TODO  should this throw for any chain id not one of celo's chains or is that to restrictive?
+  if (chainId <= 0) throw new InvalidChainIdError({ chainId })
+  if (to && !isAddress(to)) throw new InvalidAddressError({ address: to })
+  if (gasPrice)
+    throw new BaseError(
+      '`gasPrice` is not a valid CIP-42 Transaction attribute.',
+    )
+  if (maxFeePerGas && maxFeePerGas > 2n ** 256n - 1n)
```
Comment (aaronmgdr) [inline src/chains/serializers/celo.ts:135]: using same logic as for EIP1559 -- need to double check

File: src/chains/serializers/celo.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  return concatHex([
+    '0x7c',
+    toRlp(serializedTransaction),
+  ]) as SerializedCIP42TransactionReturnType
+}
+
+// process as CIP42 if any of these fields are present realistically gatewayfee is not used but is part of spec
+function couldBeCIP42(tx: TransactionSerializableIncludingCIP42) {
+  const maybeCIP42 = tx as TransactionSerializableCIP42
+  if (
```
Comment (aaronmgdr) [inline src/chains/serializers/celo.ts:114]: must have maxGasFee

**Comment by jxom** [comment]
Looking great! Just need to get some coverage on these lines: 
Screenshot 2023-06-30 at 7 30 56 am

**Comment by aaronmgdr** [comment]
> Looking great! Just need to get some coverage on these lines: Screenshot 2023-06-30 at 7 30 56 am

Ive added test for the lines that were showen as missing. but on CI its saying almost none of the file is covered. :thinking_face:  This is odd since the lines it points to are the entire serializeTransaction and serializeTransactionCIP42 functions. 


Also I create a PR on top of the first that creates some transaction argument assertion functions. I thought i might be useful to have these in one place rather than rewriting for each serializer. 

https://github.com/clabs-co/viem/pull/2/files

**Review by jxom — APPROVED** [review]

**Pull request merged**